Weekly_Work/Week1/Code_changes-git/Metrics_Root_class.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+import json
+import datetime
+import matplotlib.pyplot as plt
+import pandas as pd
+
+
+# In[7]:
+
+
+class Metric():
+    
+    def __init__(self, path, since=None, until=None, source_code_exclude_list=None):
+
+        self.since = since
+        self.until = until
+        self.source_code_exclude_list = source_code_exclude_list
+        self.repo_urls = set()
+        
+        clean_commit_list = list()
+        clean_issue_list = list()
+        clean_pr_list = list()
+        
+        with open(path, 'r') as raw_data:
```

Comment:
I don't understand this code well... I think we should have a clean function for initializing with data from the JSON file, and then, maybe, some other functions for filtering what was read (and these would likely be different for the Pandas and the list of dictionaries case).

In fact, it would be better that the loading code just produces some in-memory data structure (a Pandas dataframe in one case, a list of dictionaries in the other), and then this data structure is what is used to initialize the class. This would allow to easily read once from the file, and then instantiating several objects, from the corresponding classes, using the same data. This could also help to interface directly to Perceval as well (not useful for the reference implementations, but maybe for later code).

This way, the first lines of every reference implementation could be something like:

```
df = R_df.read(filename)
metrics_object = M_df.metrics_class(df, ...)
```

or

```
list = R_l.read(filename)
metrics_object = M_l.metrics_class(list, ...)
```

In these examples, R_df would be the "root" module for dataframes, with a function to read from JSON files, producing dataframes (`read`), and the root class for metrics computed based on dataframes. M_df would be the module implementing a certain metric M, given a dataframe, with a class metrics_class 'metrics_class' (child of `R_df.root_class`). `_l` would be the same for lists of dictionaries.

This way, we would separate reading the JSON files from the hierarchy of metrics, and implementations with lists of dictionaries from implementations with dataframes.

Reply:
> I don't understand this code well...
```python
clean_commit_list = list()
clean_issue_list = list()
clean_pr_list = list()
```
My code does not mix both implementations exactly, it is just that it is more convenient to populate a dataframe with a list of dictionaries, where each element of that list will end up being a row in a DataFrame.

Thus, those lists mentioned above have commits, issues and pull requests as their elements and then the dataframe can be populated using these lists.

Regarding the structure and after reading your comments, we could have a structure like this:
```
Root / 
 - Root_df.py (The Root class for dataframe implementation)
 - Root_list.py ( The root class for list implementation)

Metrics_df /
  - each_metric.py (One metric class per file dataframe implementation)
  - ..
  - ..

Metrics_list /
  - each_metric_list.py (One metric class per file, list implementation)
  - ..
  - ..
```
> This way, we would separate reading the JSON files from the hierarchy of metrics,

I don't think I understand what you mean by this. The reading and cleaning of JSON file (which is currently happening in the __init__ method of the root class) happens away from the metrics. If I have misunderstood, please help me out here a bit!
Thanks.

Reply:
> Thus, those lists mentioned above have commits, issues and pull requests as their elements and then the dataframe can be populated using these lists.

Maybe there are exceptions, but usually, the metrics we're computing are on commits, pull requests *or* issues. That means that the corresponding dataframe will have only one of those. Probably that's the reason why i didn't understood the code. I was envisioning code in the root class just to read the data from JSON files, without further processing. It would be specialized code in, say, chile classes for computing metrics related to commits, the code that will do something with commits (and only commits).

Reply:
> we could have a structure like this

I think that's what I have in mind, but I'm not completely sure. What is `Root /`? If that's a directory, I think we don't need it (for now). If it something else, please let me know what it is (same for `Metrics_df /`, for example).

Reply:
Thanks for the prompt reply!
So each metric would end up having two implementations and thus, probably two root classes (one for pandas, and the other for a non-pandas implementation). So, the `Root` directory would have the two root classes and `Metrics_df` would be a directory which would have all the pandas style reference implementation files. Similarly, for Metrics_list.

> If that's a directory, I think we don't need it

Then what structure do you have in mind?

Reply:
For now, I think we can just have all implementations in the same directory. Later on, maybe we can use subdirectories or something else, but I'm still not sure if that's needed. For example, for metrics definitions, for now we have a flat structure in a single directory.